feat: Добавлена новая диагностика `ServerCallsInFormEventsDiagnostic`

johnnyshut opened the pull request:
## Описание

Добавлена новая диагностика `ServerCallsInFormEventsDiagnostic` для обнаружения серверных вызовов в событиях `ПриАктивизацииСтроки` (OnActivateRow) и `НачалоВыбора` (OnStartChoice) в модулях форм.

### Внесенные изменения:

1. **Новая диагностика** - `ServerCallsInFormEventsDiagnostic.java`
   - Обнаруживает вызовы серверных процедур в клиентских событиях форм
   - Работает только в модулях типа `FormModule`
   - Поддерживает как русские, так и английские названия событий

2. **Локализация** — добавлены файлы сообщений на русском и английском языках
   - `ServerCallsInFormEventsDiagnostic_ru.properties`
   - `ServerCallsInFormEventsDiagnostic_en.properties`

3. **Документация** — создана документация для диагностики
   - `docs/diagnostics/ServerCallsInFormEvents.md` (русский)
   - `docs/en/diagnostics/ServerCallsInFormEvents.md` (английский)
   - Включает описание проблемы, примеры и ссылку на статью Infostart

4. **Тесты** — добавлены comprehensive unit тесты
   - `ServerCallsInFormEventsDiagnosticTest.java` - unit тесты
   - `ServerCallsInFormEventsDiagnostic.bsl` - BSL тестовый файл
   - Покрывают все сценарии использования

### Технические детали:
- Диагностика использует visitor pattern для обхода AST
- Проверяет тип модуля перед выполнением
- Отслеживает состояние входа/выхода из запрещенных событий
- Срабатывает только на вызовы серверных процедур в клиентских событиях

## Связанные задачи
<!--- Для каждого PR обязательно наличие связанной задачи (issue). -->
<!--- Необходимо указать ключи задач, предваряя их символом #, например -->
<!---Closes #123 -->
<!--  -->
<!-- ВНИМАНИЕ: Без ссылки на задачу пулл-реквест не будет принят! -->
<!--  -->
Closes #3473

## Чеклист
<!--- Перед отправкой пройдите по списку и поставьте отметку для каждого выполненного действия -->
<!--- Если не понятно, что подразумевается — спросите в чате проекта -->

### Общие

- [x] Ветка PR обновлена из develop
- [x] Отладочные, закомментированные и прочие, не имеющие смысла участки кода удалены
- [x] Изменения покрыты тестами
- [x] Обязательные действия перед коммитом выполнены (запускал команду `gradlew precommit`)

### Для диагностик

- [x] Описание диагностики заполнено для обоих языков (присутствуют файлы для обоих языков, для русского заполнено все подробно, перевод на английский можно опустить)

## Дополнительно
<!--- Различная дополнительная информация, скриншоты и т.д. -->

### Ссылки на источники:
- [Статья Infostart](https://infostart.ru/1c/articles/1225834/) - источник требований к диагностике

### Примеры использования:

**Неправильно (вызовет диагностику):**
```bsl
Процедура ПриАктивизацииСтроки(Элемент, ВыбраннаяСтрока, Поле, НовоеЗначение, СтандартнаяОбработка)
    // Ошибка: серверный вызов в клиентском событии
    ТаблицаФормыПриАктивизацииСтрокиНаСервере();
    СтандартнаяОбработка = Ложь;
КонецПроцедуры
```

**Правильно:**
```bsl
Процедура ПриАктивизацииСтроки(Элемент, ВыбраннаяСтрока, Поле, НовоеЗначение, СтандартнаяОбработка)
    // Правильно: только клиентская логика
    СтандартнаяОбработка = Ложь;
КонецПроцедуры
```

### Тестирование:
- Все unit тесты проходят успешно
- Диагностика корректно работает только в модулях форм
- Поддерживает как русские, так и английские названия событий

---

**Примечание:** Перед отправкой PR необходимо выполнить команду `gradlew precommit` для проверки всех обязательных действий.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Diagnostic that flags server procedure calls inside client form events (OnActivateRow / OnStartChoice and Russian equivalents).

* **Documentation**
  * Added RU/EN docs describing the diagnostic with incorrect vs correct examples and references.

* **Tests**
  * Added comprehensive unit tests covering Russian/English names, case-insensitivity, edge cases, multiple calls and directive handling.

* **Configuration**
  * New schema/parameter to enable/adjust the diagnostic.

* **Localization**
  * Added English and Russian diagnostic messages.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

theshadowco commented on docs/diagnostics/ServerCallsInFormEvents.md:
```diff
@@ -0,0 +1,43 @@
+## Описание диагностики
+<!-- Описание диагностики заполняется вручную. Необходимо понятным языком описать смысл и схему работу -->
+
+В событиях `ПриАктивизацииСтроки` и `НачалоВыбора` не должно быть вызовов серверных процедур. Эти события должны выполняться только на клиенте.
```
стоит добавить, что диагностируются только серверные вызовы методов ЭТОЙ формы, т.е. вызов `МойМодульСервер.МойСерверныйМетод` не будет ошибкой

theshadowco commented on src/main/resources/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic_ru.properties:
```diff
@@ -0,0 +1,11 @@
+# Сообщения диагностики ServerCallsInFormEventsDiagnostic на русском языке
+
+# Название диагностики
```
см выше

johnnyshut commented:
@theshadowco исправил замечания

nixel2007 commented on src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java:
```diff
@@ -0,0 +1,157 @@
+   * @return результат обхода дерева
+   */
+  @Override
+  public ParseTree visitSub(BSLParser.SubContext ctx) {
```
1. Здесь больше подходит AbstractListenerDiagnostic, так как тебе надо сбрасывать флаги на выходе.
2. Для неподходящих процедур нет смысла дёргать super, это будет в холостую гонять визитор по дереву. Для таких случаев надо сразу возвращать ctx.
3. Вместо разбора дерева выражений (GlobalMethodCall) можно через ReferenceIndex посмотреть все исходящие ссылки из текущего метода. Тогда по дереву вглубь ходить не надо вообще. 
4. Возможно проще и быстрее будет сделать AbstractSymbolSymbolTreeDiagnostic, отвязавшись от ParseTree полностью
5. Или пойти ещё дальше, и сделать AbstractDiagnostic, получить все method symbols из symbol tree, отфильтровать нужные и их проверить. Но это уже на твоё усмотрение.

johnnyshut replied:
Посмотри теперь на реализацию. И запусти, пожалуйста, еще раз сборку, один flow упал, но вроде ошибка просто про то, что файл не удалось прочитать.

johnnyshut committed:
feat(diagnostics): добавлен ServerCallsInFormEventsDiagnostic для проверки событий формы.

- Добавлена ​​новая диагностика для предотвращения вызовов серверных процедур в событиях OnActivateRow и OnStartChoice, которые могут привести к проблемам в работе формы.
- Обновлена ​​схема конфигурации, включающая настройки для новой диагностики.
- Добавлены английская и русская локализации для диагностических сообщений.

johnnyshut committed:
test(diagnostics): добавлены тесты для ServerCallsInFormEventsDiagnostic

- Создан новый тестовый класс `ServerCallsInFormEventsDiagnosticTest` для проверки корректности обнаружения вызовов серверных процедур в событиях формы.
- Добавлены тесты на различные сценарии, включая проверки для русских и английских названий событий, регистронезависимых названий, а также отсутствие диагностик в обычных процедурах и разрешенных событиях.
- Включены тесты для проверки множественных вызовов серверных процедур и пустых обработчиков событий.
- Добавлен тестовый файл `ServerCallsInFormEventsDiagnostic.bsl` с примерами правильных и неправильных событий.

johnnyshut committed:
doc(diagnostics): добавлены документы для диагностики вызовов серверных процедур в событиях форм

- Созданы новые файлы документации на русском и английском языках для диагностики, предотвращающей вызовы серверных процедур в событиях `ПриАктивизацииСтроки` и `НачалоВыбора`.
- Описаны проблемы, которые могут возникнуть из-за таких вызовов, и приведены примеры правильного и неправильного использования.
- Добавлены ссылки на источники информации, включая статью на Infostart и GitHub Issue.

johnnyshut committed:
refactor(diagnostics): удален устаревший метод enterMethodCall из ServerCallsInFormEventsDiagnostic

- Удален метод enterMethodCall, который обрабатывал вызовы локальных методов и добавлял диагностические сообщения для серверных процедур в событиях формы.

theshadowco commented on src/main/resources/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic_ru.properties:
```diff
@@ -0,0 +1,3 @@
+diagnosticName=Серверные вызовы в событиях форм
+diagnosticMessage=В событиях ПриАктивизацииСтроки и НачалоВыбора не должно быть вызовов серверных процедур. Процедура "{0}" выполняется на сервере, что может привести к проблемам.
+message=В событиях ПриАктивизацииСтроки и НачалоВыбора не должно быть вызовов серверных процедур. Процедура "{0}" выполняется на сервере, что может привести к проблемам.
```
а зачем message? diagnosticMessage достаточно, или я что-то не вижу?

johnnyshut replied:
Использовал кастомный ключ (info.getMessage("message", methodName)), исправил на дефолтный

johnnyshut committed:
refactor(diagnostics): упрощено сообщение диагностики в ServerCallsInFormEventsDiagnostic

- Удалено устаревшее сообщение из файлов локализации для диагностики вызовов серверных процедур в событиях форм.
- Обновлено добавление диагностического сообщения, чтобы использовать только имя метода вместо полного сообщения.
